src/iqm/qiskit_iqm/iqm_provider.py

```diff
@@ -241,119 +241,148 @@ def serialize_circuit(self, circuit: QuantumCircuit) -> Circuit:
-                native_inst = Instruction(name='prx', qubits=qubit_names, args={'angle_t': angle_t, 'phase_t': 0.25})
-            elif instruction.name == 'cz':
-                native_inst = Instruction(name='cz', qubits=qubit_names, args={})
-            elif instruction.name == 'move':
-                native_inst = Instruction(name='move', qubits=qubit_names, args={})
-            elif instruction.name == 'barrier':
-                native_inst = Instruction(name='barrier', qubits=qubit_names, args={})
-            elif instruction.name == 'measure':
-                if len(circuit_instruction.clbits) != 1:
-                    raise ValueError(
-                        f'Unexpected: measurement instruction {circuit_instruction} uses multiple classical bits.'
-                    )
-                clbit = circuit_instruction.clbits[0]  # always a single-qubit measurement
-                mk = str(MeasurementKey.from_clbit(clbit, circuit))
-                native_inst = Instruction(name='measure', qubits=qubit_names, args={'key': mk})
-                clbit_to_measure[clbit] = native_inst
-            elif instruction.name == 'reset':
-                # implemented using a measure instruction to measure the qubits, and
-                # one cc_prx per qubit to conditionally flip it to |0>
-                feedback_key = '_reset'
+        instructions = serialize_instructions(circuit, self._idx_to_qb, ignore_nonnative_gates=False)
+
+        try:
+            metadata = to_json_dict(circuit.metadata)
+        except ValueError:
+            warnings.warn(
+                f'Metadata of circuit {circuit.name} was dropped because it could not be serialised to JSON.',
+            )
+            metadata = None
+
+        return Circuit(name=circuit.name, instructions=instructions, metadata=metadata)
+
+
+def serialize_instructions(
+    circuit: QuantumCircuit, index_to_qubit_mapping: dict[int, str], ignore_nonnative_gates: bool
+) -> list[Instruction]:
+    """Serialize a quantum circuit into Instructions in the IQM data transfer format.
+
+    This is an internal helper for :meth:`.IQMBackend.serialize_circuit` that gives slightly more control.
+    See :meth:`.IQMBackend.serialize_circuit` for details.
+    There is usually no need to use this function directly.
+
+    Args:
+        circuit: quantum circuit to serialize
+        index_to_qubit_mapping: Mapping from logical to physical qubit names.
+        ignore_nonnative_gates: If False (default), any instruction that cannot be serailizaed into the native
+            gate set will raise an error. If True, such instructions are converted as-is without validation,
+            and the caller must edit the result to be valid and executable.
+            Notably, since IQM transfer format requires named parameters and qiskit parameters don't have names, the
+            `i` th parameter of an unrecognized instruction is given the name ``"param<i>"``.
```

Comment:
Would it be better if this was a list of allowed gate names? Then we could retain some control over transpilation errors etc.

Reply:
I mean allowed on top of the hardcoded ones in the if/elif clauses.

Reply:
I feel it's unnecessary complication for the user.
A Pulla user already has to be explicit about registering new gates. If a gate is not registered, the will get the "Gate not registered" error from iqm-pulse.

Reply:
Yes, but some of those gates may come from Qiskit itself if they forgot to transpile the circuit first. It may not be clear to the user why they were not accepted by iqm-pulse. Whereas if we use a user-submitted whitelist, those gates will be passed but the serialization function will give a helpful error message about the others.

Reply:
Changed.